api/proto/teleport/discoveryconfig/v1/discoveryconfig_service.proto

```diff
@@ -39,6 +39,9 @@ service DiscoveryConfigService {
   // UpdateDiscoveryConfig updates an existing DiscoveryConfig resource.
   rpc UpdateDiscoveryConfig(UpdateDiscoveryConfigRequest) returns (DiscoveryConfig);
 
+  // UpsertDiscoveryConfig creates or updates a DiscoveryConfig resource.
+  rpc UpsertDiscoveryConfig(UpsertDiscoveryConfigRequest) returns (DiscoveryConfig);
```

Comment:
Will this require optimistic locking?

Reply:
I think all resources will use it.

This is being used in `tctl create -f` and cache maintenance.

Reply:
Upsert is always unconditional - we considered adding a primitive backend operation for "write if item doesn't exist or if it exists and it matches this revision" but we couldn't really think of uses for it. 🤷

Reply:
If you meant in general then yeah, unfortunately we have an implemented and released `UpdateDiscoveryConfig` that no one should use so the conditionally updating RPC used by `tctl edit` will probably be `UpdateDiscoveryConfigV2` or something like that. 🥲

Reply:
I meant the Update endpoint but github doesn't let me add the comment to the correct place

Reply:
I'm not sure I'm following here.

`tctl edit` should use the `UpdateDiscoveryConfig` which conditionally updates the resource.
It's not doing that right now, but the flow is there:
https://github.com/gravitational/teleport/blob/435e7d4f449866b3d74a795be8b2b2f6120d65e2/tool/tctl/common/edit_command.go#L159-L172